src/combinatorics/set_family/compression/uv.lean

```diff
@@ -0,0 +1,265 @@
+    { exact or.inr ⟨by rwa compress_idem, b, hb, (compress_idem _ _ _).symm⟩ } },
+  have : filter (λ a, compress u v a ∉ 𝓒 u v s) (𝓒 u v s) = ∅,
+    rw ←filter_false (𝓒 u v s), apply filter_congr, simpa,
+  rw [compression, compress_motion, this, image_empty, union_comm,
+      compress_remains, ←this],
+  exact filter_union_filter_neg_eq _ (compression u v s),
+end
+
+/-- Compressing a set family doesn't change its size. -/
+lemma card_compression (u v : α) (s : finset α) : (𝓒 u v s).card = s.card :=
+begin
+  rw [compression, card_disjoint_union (compress_disjoint _ _), compress_motion,
+    card_image_of_inj_on, ←card_disjoint_union, union_comm, compress_remains,
+    filter_union_filter_neg_eq],
+  { rw [disjoint.comm, disjoint_iff_inter_eq_empty],
+    exact filter_inter_filter_neg_eq _ _ },
+  intros a ha b hb hab,
+  dsimp at hab,
+  rw [mem_coe, mem_filter] at ha hb,
+  rw compress at ha hab,
+  split_ifs at ha hab with has,
+  { rw compress at hb hab,
+    split_ifs at hb hab with hbs,
+    { have h : (a ⊔ u) \ v \ u ⊔ v = (b ⊔ u) \ v \ u ⊔ v,
+      { rw hab },
+      rwa [sdiff_sdiff_comm, has.1.symm.sup_sdiff_cancel_right,
+          sdiff_sdiff_comm, hbs.1.symm.sup_sdiff_cancel_right,
+          sdiff_sup_cancel has.2, sdiff_sup_cancel hbs.2] at h },
+    { exact (hb.2 hb.1).elim } },
+  { exact (ha.2 ha.1).elim }
+end
+
+/-- If `a` is in the compressed family and does move under compression, then the compressed version
+was in the original family. -/
+lemma uncompressed_was_already_there (ha : a ∈ 𝓒 u v s) (hva : v ≤ a) (hua : disjoint u a) :
+  (a ⊔ u) \ v ∈ s :=
+begin
+  rw [mem_compression, compress_of_disjoint_of_le hua hva] at ha,
+  obtain ⟨_, ha⟩ | ⟨_, b, hb, rfl⟩ := ha,
+  { exact ha },
+  have hu : u = ⊥,
+  { suffices : disjoint u (u \ v),
+    { rwa [(hua.mono_right hva).sdiff_eq_left, disjoint_self] at this },
+    refine hua.mono_right _,
+    rw [←compress_idem, compress_of_disjoint_of_le hua hva],
+    exact sdiff_le_sdiff_right le_sup_right },
+  have hv : v = ⊥,
+  { rw ←disjoint_self,
+    apply disjoint.mono_right hva,
+    rw [←compress_idem, compress_of_disjoint_of_le hua hva],
+    exact disjoint_sdiff_self_right },
+  rwa [hu, hv, compress_self, sup_bot_eq, sdiff_bot],
+end
+
+end generalized_boolean_algebra
+
+/-! ### UV-compression on finsets -/
+
+open_locale finset_family
+
+variables [decidable_eq α] {𝒜 : finset (finset α)} {U V A : finset α}
+
+/-- Compressing a finset doesn't change its size. -/
+lemma card_compress (hUV : U.card = V.card) (A : finset α) : (compress U V A).card = A.card :=
+begin
+  unfold compress,
+  split_ifs,
+  { rw [card_sdiff (h.2.trans (le_sup_left)), sup_eq_union, card_disjoint_union h.1.symm, hUV,
+    add_tsub_cancel_right] },
+  { refl }
+end
+
+/-- If `A` is in the compressed family but `V` is a subset of `A`, `A` must have been in the
+original family. -/
+lemma compress_held (hA : A ∈ 𝓒 U V 𝒜) (hVA : V ⊆ A) (hVU : V = ∅ → U = ∅) : A ∈ 𝒜 :=
```

Comment:
hVU seems like a weird constraint

Reply:
It originally was `V.card = U.card`, but the proof only used that it implies `V = ∅ → U = ∅` so I weakened it.